In [39]:
import os
import sys
import shutil
import pandas
import numpy
from ipywidgets import interact
import ipystata
from ipystata.config import config_stata
config_stata('/usr/local/stata/stata-mp')

from scipy import stats
import statsmodels.api as sm

# import plotly
# from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot
# from plotly.graph_objs import Scatter, Figure, Layout, Histogram
# init_notebook_mode(connected=True)
# import PrettyTable as pt
# import beakerx

pandas.set_option('display.float_format', lambda x: '%.2f' % x)
pandas.set_option('max_columns', 200)
pandas.set_option('max_rows', 400)
pandas.set_option('max_colwidth', 150)
pandas.set_option('mode.sim_interactive', True)
pandas.set_option('colheader_justify', 'center')

from IPython.display import display, HTML, Image, IFrame
from IPython.html.widgets import interact, fixed

# from IPython.core.interactiveshell import InteractiveShell
# InteractiveShell.ast_node_interactivity = "all"
# pandas.options.display.float_format = '{:.2f}'.format

In [40]:
d1 = pandas.read_csv('./data/IzzyLongitudinalWorkingMarch2015.csv')
d2 = pandas.read_csv('./data/AnalysisObservations.csv')

In [41]:
d1.head()

ID   Session  Sequence  Gender   Age  Height  Seated Height  Leg Length  \
0  100     1         1        0   11.75  152.00     75.20         76.80      
1  100     1         3        0   12.71  160.80     79.60         81.20      
2  100     1         5        0   13.83  173.30     87.60         85.70      
3  100     1         7        0   14.74  177.40     89.80         87.60      
4  101     1         1        0   11.45  158.10     81.00         77.10      

   Weight  BF_Slaughter  BIA_PBF   BMI  LBM (kg)  Years from aPHV  Age_At_PHV  \
0  44.20      22.81      21.90   19.13   34.12        -2.13          13.88      
1  48.80      17.24      14.90   18.87   40.39        -1.21          13.92      
2  63.50      21.69      17.60   21.14   49.73         0.41          13.42      
3  67.30      18.60      13.10   21.38   54.78         1.20          13.57      
4  47.30      23.71      20.70   18.92   36.08        -1.63          13.08      

   Mat_Offset_New  Bone_Age  Tanner  Menarche Age_First_Period  \
0      -1.67        11.16     2.00     nan           NaN         
1      -0.71        12.41     4.00     nan           NaN         
2       0.41        14.66     4.00     nan           NaN         
3       1.17        15.66     5.00     nan           NaN         
4      -1.63        12.24     1.00     nan           NaN         

  Period_Frequency  Medications  Vitamins  Fracture  Radius_SOS  Radius_Z  \
0        NaN           0.00        0.00        0      3828.00      0.80     
1        NaN           0.00        0.00        0      3898.00      1.00     
2        NaN           0.00        0.00        0      3851.00      0.50     
3        NaN           0.00        0.00        0      3952.00      1.30     
4        NaN           0.00        0.00        0      3682.00     -1.10     

   Tibial_SOS  Tibial_Z  Wrist_Flexion_Best  Grip_Strength_Best  \
0   3601.00     -0.75            nan                 nan          
1   3629.00     -0.70          13.65               27.00          
2   3677.00      0.30          19.25               37.00          
3   3740.00      0.30          24.40               40.50          
4   3603.00     -0.90            nan                 nan          

   Forearm_Length  ForeArm_Lean_CSA  Forearm_US_CSA  Anterior_CSA  \
0        nan              nan              nan            nan       
1      25.60            30.41            14.49          11.97       
2      27.70            39.61            19.83          16.30       
3      28.50            43.97            20.97          17.52       
4        nan              nan              nan            nan       

   Anterior Radius  Ant_Rad_CSA  Anterior Ulna  Ant_Ulna_CSA  Posterior  \
0        nan            nan           nan            nan         nan      
1      34.74           9.48         17.81           2.49       17.90      
2      41.07          13.25         19.70           3.05       21.23      
3      42.15          13.95         21.32           3.57       20.97      
4        nan            nan           nan            nan         nan      

   Post_CSA  Tibialis Anterior    NTx   Creatinine  NTx/Creat  Caloric Intake  \
0     nan          nan         1443.05     2.03      711.85       1572.00       
1    2.52          nan         3738.62     4.92      760.09       2219.01       
2    3.54          nan         4027.06     7.41      543.73       2599.83       
3    3.45          nan         2631.34     5.79      454.36       2482.12       
4     nan          nan         9869.88    10.52      938.00       1549.99       

   Calcium  Vit D  Vit D.1  Caffeine  tot sedetary  tot light  tot mod  \
0 1077.67   3.38   135.27    16.10       78.43       174.29    94.57     
1 1212.82   1.02    41.02     2.14       75.55       148.71    48.79     
2  874.34   4.55   181.85     2.25       78.43       185.57    88.00     
3 1672.30   5.91   236.74     6.19       75.14       156.57    73.71     
4 1521.05   7.43   297.30     6.99       65.57       174.71  

In [42]:
d2.head()

id   season  sequence gender  rsos  tsos  grip  ptiso  Biodex  MatLab  \
0  100  spring      1      boy   3828  3601   nan 113.93  129.10  113.93   
1  100  spring      3      boy   3898  3629 27.00 136.02  138.70  136.02   
2  100  spring      5      boy   3851  3677 37.00 177.05  178.00  177.05   
3  100  spring      7      boy   3952  3740 40.50 205.80  205.80  205.80   
4  101  spring      1      boy   3682  3603   nan 133.05  139.10  133.05   

   ntxc   matu   mvh    age Exclusion  NoSOS         NoGrip        Noptiso  \
0 711.85 -1.67 105.71 11.75    none    nan    Not taken sequence 1    NaN    
1 760.09 -0.71  93.62 12.71    none    nan                     NaN    NaN    
2 543.73  0.41  98.14 13.83    none    nan                     NaN    NaN    
3 454.36  1.17  88.14 14.74    none    nan                     NaN    NaN    
4 938.00 -1.63  89.29 11.45    none    nan    Not taken sequence 1    NaN    

  Nontx Nomvh  T    _o  _id  _v  size _season  omit  
0   NaN   NaN  NaN   1   1    1  4.00  spring  nan   
1   NaN   NaN  NaN   2   1    2   nan     NaN  nan   
2   NaN   NaN  NaN   3   1    3   nan     NaN  nan   
3   NaN   NaN  NaN   4   1    4   nan     NaN  nan   
4   NaN   NaN  NaN   5   2    1  1.00  spring  nan

In [43]:
dt = d1[['ID', 'Sequence', 'Caloric Intake', 'Height', 'Weight']]
DM = pandas.merge(dt, d2, left_on=['ID', 'Sequence'], right_on=['id', 'sequence'], how='right'); del dt
DM.head()

ID   Sequence  Caloric Intake  Height  Weight  id   season  sequence  \
0  100      1        1572.00      152.00  44.20   100  spring      1      
1  100      3        2219.01      160.80  48.80   100  spring      3      
2  100      5        2599.83      173.30  63.50   100  spring      5      
3  100      7        2482.12      177.40  67.30   100  spring      7      
4  101      1        1549.99      158.10  47.30   101  spring      1      

  gender  rsos  tsos  grip  ptiso  Biodex  MatLab  ntxc   matu   mvh    age  \
0   boy   3828  3601   nan 113.93  129.10  113.93 711.85 -1.67 105.71 11.75   
1   boy   3898  3629 27.00 136.02  138.70  136.02 760.09 -0.71  93.62 12.71   
2   boy   3851  3677 37.00 177.05  178.00  177.05 543.73  0.41  98.14 13.83   
3   boy   3952  3740 40.50 205.80  205.80  205.80 454.36  1.17  88.14 14.74   
4   boy   3682  3603   nan 133.05  139.10  133.05 938.00 -1.63  89.29 11.45   

  Exclusion  NoSOS         NoGrip        Noptiso Nontx Nomvh  T    _o  _id  \
0    none    nan    Not taken sequence 1    NaN    NaN   NaN  NaN   1   1    
1    none    nan                     NaN    NaN    NaN   NaN  NaN   2   1    
2    none    nan                     NaN    NaN    NaN   NaN  NaN   3   1    
3    none    nan                     NaN    NaN    NaN   NaN  NaN   4   1    
4    none    nan    Not taken sequence 1    NaN    NaN   NaN  NaN   5   2    

   _v  size _season  omit  
0   1  4.00  spring  nan   
1   2   nan     NaN  nan   
2   3   nan     NaN  nan   
3   4   nan     NaN  nan   
4   1  1.00  spring  nan

In [44]:
dm = DM.copy()
dm = dm.sort_values(by=['id', 'sequence'])

dm['boys'] = numpy.where(dm.gender=='boy', 1, 0)
dm['girls'] = 1 - dm.boys


dm['trip'] = dm.groupby(['id']).cumcount()+1
dm = dm.rename(columns={'sequence': 'session'})
dm = dm.rename(columns={'Biodex': 'biodex'})
dm = dm.rename(columns={'MatLab': 'matlab'})
dm = dm.rename(columns={'Caloric Intake': 'calo'})
dm = dm.rename(columns={'Height': 'height'})
dm = dm.rename(columns={'Weight': 'weight'})

dm = dm[['id', 'trip', 'session', 'season', 
         'rsos', 'tsos', 
         'grip', 'ptiso','biodex', 'matlab', 
         'ntxc', 'calo', 'mvh', 
         'matu', 'age', 'gender', 'boys', 'girls', 'height', 'weight',
         'Exclusion', 'NoSOS', 'NoGrip', 'Noptiso', 'Nontx', 'Nomvh', 'T']]
dm.head()

id   trip  session  season  rsos  tsos  grip  ptiso  biodex  matlab  ntxc   \
0  100    1      1     spring  3828  3601   nan 113.93  129.10  113.93 711.85   
1  100    2      3     spring  3898  3629 27.00 136.02  138.70  136.02 760.09   
2  100    3      5     spring  3851  3677 37.00 177.05  178.00  177.05 543.73   
3  100    4      7     spring  3952  3740 40.50 205.80  205.80  205.80 454.36   
4  101    1      1     spring  3682  3603   nan 133.05  139.10  133.05 938.00   

    calo    mvh   matu   age gender  boys  girls  height  weight Exclusion  \
0 1572.00 105.71 -1.67 11.75   boy     1     0    152.00  44.20     none     
1 2219.01  93.62 -0.71 12.71   boy     1     0    160.80  48.80     none     
2 2599.83  98.14  0.41 13.83   boy     1     0    173.30  63.50     none     
3 2482.12  88.14  1.17 14.74   boy     1     0    177.40  67.30     none     
4 1549.99  89.29 -1.63 11.45   boy     1     0    158.10  47.30     none     

   NoSOS         NoGrip        Noptiso Nontx Nomvh  T    
0  nan    Not taken sequence 1    NaN    NaN   NaN  NaN  
1  nan                     NaN    NaN    NaN   NaN  NaN  
2  nan                     NaN    NaN    NaN   NaN  NaN  
3  nan                     NaN    NaN    NaN   NaN  NaN  
4  nan    Not taken sequence 1    NaN    NaN   NaN  NaN

In [45]:
dm.describe()

id    trip   session   rsos    tsos   grip   ptiso  biodex  matlab  \
count 406.00 406.00  406.00   406.00  406.00 363.00 345.00  406.00  345.00   
mean  333.04   1.86    4.19  3816.08 3687.41  23.43 128.04  131.48  128.04   
std   197.62   0.88    1.83   100.01  109.42   7.40  56.68   57.73   56.68   
min   100.00   1.00    1.00  3554.00 3367.00   9.00  32.65   41.20   32.65   
25%   140.00   1.00    3.00  3747.50 3619.50  18.00  86.12   88.84   86.12   
50%   194.00   2.00    4.00  3820.00 3688.00  22.00 115.25  118.65  115.25   
75%   531.75   2.00    6.00  3878.00 3757.75  27.00 160.50  161.07  160.50   
max   583.00   4.00    7.00  4140.00 4015.00  53.00 353.00  353.80  353.00   

        ntxc    calo    mvh   matu    age   boys   girls  height  weight  \
count  376.00  404.00 298.00 406.00 406.00 406.00 406.00  406.00  406.00   
mean   539.82 1593.43 107.05  -0.86  11.84   0.51   0.49  152.15   46.12   
std    263.65  479.51  41.53   2.02   1.99   0.50   0.50   13.61   14.43   
min     56.31  469.50  13.57  -5.30   7.50   0.00   0.00  122.80   20.50   
25%    370.82 1265.06  79.47  -2.29  10.34   0.00   0.00  141.90   35.23   
50%    514.00 1541.21  99.53  -0.99  11.84   1.00   0.00  152.00   44.65   
75%    662.49 1848.58 134.03   0.52  13.23   1.00   1.00  162.17   54.00   
max   1754.32 3624.72 235.00   9.66  17.55   1.00   1.00  187.60  116.90   

       NoSOS  
count  0.00   
mean    nan   
std     nan   
min     nan   
25%     nan   
50%     nan   
75%     nan   
max     nan

In [46]:
ds = dm.groupby('session')['rsos', 'tsos', 'grip', 'ptiso','biodex', 'matlab', 'ntxc', 'calo', 'matu', 'mvh', 'age', 'gender'].mean()
d0 = dm[dm['gender']==0].groupby('session')['rsos', 'tsos', 'grip', 'ptiso','biodex', 'matlab', 'ntxc', 'calo', 'matu', 'mvh', 'age'].mean()
d1 = dm[dm['gender']==1].groupby('session')['rsos', 'tsos', 'grip', 'ptiso','biodex', 'matlab', 'ntxc', 'calo', 'matu', 'mvh', 'age'].mean()

t1 = ds.append(d0)
t1 = t1.append(d1)
t1

rsos    tsos   grip  ptiso  biodex  matlab  ntxc    calo   matu  \
session                                                                     
1       3798.51 3641.84   nan 109.12  111.64  109.12 741.95 1574.72 -1.85   
2       3771.03 3657.03 19.13  78.98  101.14   78.98 567.99 1401.74 -1.88   
3       3790.04 3656.50 20.17 114.10  120.07  114.10 555.63 1639.72 -1.17   
4       3819.89 3683.83 22.66 127.32  130.79  127.32 527.80 1535.78 -0.94   
5       3806.17 3681.09 25.15 136.45  142.14  136.45 496.39 1625.33 -0.53   
6       3858.58 3728.35 24.78 139.68  141.58  139.68 478.24 1559.54 -0.50   
7       3851.87 3753.53 28.29 161.20  160.05  161.20 462.20 1758.93  0.40   

          mvh    age  
session               
1       119.94 11.15  
2       109.12 10.90  
3       121.47 11.29  
4        90.29 11.82  
5       102.19 12.09  
6       112.63 12.25  
7        86.40 13.12

In [47]:
def get_st(var):
    temp = 'st_' + var
    dm[temp] = (dm[var] - dm[var].mean()) / dm[var].std()

_vars = ['rsos', 'tsos', 'grip', 'ptiso','biodex', 'matlab', 'ntxc', 'calo', 'matu', 'mvh', 'age', 'height', 'weight']
for v in _vars:
    get_st(v)
    
dm.head()

id   trip  session  season  rsos  tsos  grip  ptiso  biodex  matlab  ntxc   \
0  100    1      1     spring  3828  3601   nan 113.93  129.10  113.93 711.85   
1  100    2      3     spring  3898  3629 27.00 136.02  138.70  136.02 760.09   
2  100    3      5     spring  3851  3677 37.00 177.05  178.00  177.05 543.73   
3  100    4      7     spring  3952  3740 40.50 205.80  205.80  205.80 454.36   
4  101    1      1     spring  3682  3603   nan 133.05  139.10  133.05 938.00   

    calo    mvh   matu   age gender  boys  girls  height  weight Exclusion  \
0 1572.00 105.71 -1.67 11.75   boy     1     0    152.00  44.20     none     
1 2219.01  93.62 -0.71 12.71   boy     1     0    160.80  48.80     none     
2 2599.83  98.14  0.41 13.83   boy     1     0    173.30  63.50     none     
3 2482.12  88.14  1.17 14.74   boy     1     0    177.40  67.30     none     
4 1549.99  89.29 -1.63 11.45   boy     1     0    158.10  47.30     none     

   NoSOS         NoGrip        Noptiso Nontx Nomvh  T    st_rsos  st_tsos  \
0  nan    Not taken sequence 1    NaN    NaN   NaN  NaN   0.12    -0.79     
1  nan                     NaN    NaN    NaN   NaN  NaN   0.82    -0.53     
2  nan                     NaN    NaN    NaN   NaN  NaN   0.35    -0.10     
3  nan                     NaN    NaN    NaN   NaN  NaN   1.36     0.48     
4  nan    Not taken sequence 1    NaN    NaN   NaN  NaN  -1.34    -0.77     

   st_grip  st_ptiso  st_biodex  st_matlab  st_ntxc  st_calo  st_matu  st_mvh  \
0    nan     -0.25     -0.04      -0.25      0.65    -0.04    -0.40    -0.03    
1   0.48      0.14      0.13       0.14      0.84     1.30     0.07    -0.32    
2   1.83      0.86      0.81       0.86      0.01     2.10     0.63    -0.21    
3   2.31      1.37      1.29       1.37     -0.32     1.85     1.00    -0.46    
4    nan      0.09      0.13       0.09      1.51    -0.09    -0.38    -0.43    

   st_age  st_height  st_weight  
0  -0.05    -0.01      -0.13     
1   0.44     0.64       0.19     
2   1.00     1.55       1.20     
3   1.46     1.85       1.47     
4  -0.20     0.44       0.08

In [23]:
dm.to_csv('out/data_analysis.csv', index=False)

In [48]:
# pandas.set_option('display.float_format', lambda x: '%.2f' % x)
de = dm[['id', 'trip', 
         'rsos', 'tsos', 
         'grip', 'biodex', 'ntxc', 'calo', 'mvh', 
         'matu', 'gender', 'boys', 'girls']]

dt = de.copy()
dt = dt.drop(columns=['gender'])
dt['gender'] = 'All'

ds = de.copy()
ds = dt.append(ds)

ds = ds.set_index('gender')
ds = ds.rename(columns={'rsos': 'Radial SOS'})
ds = ds.rename(columns={'tsos': 'tibial SOS'})
ds = ds.rename(columns={'grip': 'grip strength'})
ds = ds.rename(columns={'biodex': 'knee extensors'})
ds = ds.rename(columns={'matu': 'maturity offset'})
ds = ds.rename(columns={'ntxc': 'ntx creatine'})
ds = ds.rename(columns={'calo': 'caloric intake'})
ds = ds.rename(columns={'mvh': 'MVH'})

dt = ds.groupby('gender').agg(['mean','std', 'count'])
dt = dt.rename(columns={'mean':'Mean'})
dt = dt.rename(columns={'std':'Standard deviation'})
dt = dt.rename(columns={'count':'Observations'})

dt.index.name = ''
dt = dt.T

dt.to_csv('table1.csv'); del ds
dt




All     boy    girl 
knee extensors  Mean                131.48  139.10  123.39
                Standard deviation   57.73   65.70   46.66
                Observations        406.00  209.00  197.00
boys            Mean                  0.51    1.00    0.00
                Standard deviation    0.50    0.00    0.00
                Observations        406.00  209.00  197.00
caloric intake  Mean               1593.43 1724.11 1453.37
                Standard deviation  479.51  503.08  410.09
                Observations        404.00  209.00  195.00
girls           Mean                  0.49    0.00    1.00
                Standard deviation    0.50    0.00    0.00
                Observations        406.00  209.00  197.00
grip strength   Mean                 23.43   24.90   21.90
                Standard deviation    7.40    8.55    5.61
                Observations        363.00  185.00  178.00
id              Mean                333.04  142.99  534.68
                Standard deviation  197.62   27.19   23.05
                Observations        406.00  209.00  197.00
maturity offset Mean                 -0.86   -1.61   -0.06
                Standard deviation    2.02    1.87    1.87
                Observations        406.00  209.00  197.00
MVH             Mean                107.05  114.86   99.64
                Standard deviation   41.53   42.59   39.22
                Observations        298.00  145.00  153.00
ntx creatine    Mean                539.82  544.56  534.83
                Standard deviation  263.65  230.29  295.34
                Observations        376.00  193.00  183.00
Radial SOS      Mean               3816.08 3806.62 3826.12
                Standard deviation  100.01   96.95  102.45
                Observations        406.00  209.00  197.00
trip            Mean                  1.86    1.82    1.90
                Standard deviation    0.88    0.86    0.90
                Observations        406.00  209.00  197.00
tibial SOS      Mean               3687.41 3676.46 3699.03
                Standard deviation  109.42  101.32  116.54
                Observations        406.00  209.00  197.00

In [49]:
%%stata --data de 

xtset id trip
xtdescribe


       panel variable:  id (unbalanced)
        time variable:  trip, 1 to 4
                delta:  1 unit

      id:  100, 101, ..., 583                                n =        173
    trip:  1, 2, ..., 4                                      T =          4
           Delta(trip) = 1 unit
           Span(trip)  = 4 periods
           (id*trip uniquely identifies each observation)

Distribution of T_i:   min      5%     25%       50%       75%     95%     max
                         1       1       2         2         3       4       4

     Freq.  Percent    Cum. |  Pattern
 ---------------------------+---------
       65     37.57   37.57 |  111.
       52     30.06   67.63 |  11..
       39     22.54   90.17 |  1...
       17      9.83  100.00 |  1111
 ---------------------------+---------
      173    100.00         |  XXXX


In [70]:
%%stata --data de

xtset id trip
xtsum rsos tsos
swilk rsos tsos
ttest rsos == tsos

*twoway (hist rsos, frac lcolor(gs12) fcolor(gs12)) (hist tsos, frac fcolor(none) lcolor(red)), legend(off) xtitle("rsos (red: tsos)")



       panel variable:  id (unbalanced)
        time variable:  trip, 1 to 4
                delta:  1 unit

Variable         |      Mean   Std. Dev.       Min        Max |    Observations
-----------------+--------------------------------------------+----------------
rsos     overall |  3816.084   100.0084       3554       4140 |     N =     406
         between |             86.03224       3622       4119 |     n =     173
         within  |             51.45185   3496.084   4012.084 | T-bar = 2.34682
                 |                                            |
tsos     overall |  3687.411   109.4166       3367       4015 |     N =     406
         between |             96.74594       3395     3925.5 |     n =     173
         within  |             49.81649   3501.161   3861.161 | T-bar = 2.34682

                   Shapiro-Wilk W test for normal data

    Variable |        Obs       W           V         z       Prob>z
-------------+----------------------------------------------

In [73]:
%%stata --data de

xtset id trip

bys gender: xtsum rsos tsos
di "radial SOS****************"
ttest rsos, by(gender)

di "tibial SOS****************"

ttest tsos, by(gender)



       panel variable:  id (unbalanced)
        time variable:  trip, 1 to 4
                delta:  1 unit

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-> gender = boy

Variable         |      Mean   Std. Dev.       Min        Max |    Observations
-----------------+--------------------------------------------+----------------
rsos     overall |  3806.622   96.94986       3554       4074 |     N =     209
         between |             83.66979       3641       4004 |     n =      93
         within  |              45.0813   3486.622   4002.622 | T-bar = 2.24731
                 |                                            |
tsos     overall |  3676.459   101.3152       3367       4015 |     N =     209
         between |             93.20911       3395     3925.5 |     n =      93
         within  |             40.71837   3523.4

In [67]:
# %%stata --data de

# xtset id trip
# xtsum grip biodex ntxc calo mvh matu 
# xtsum grip biodex ntxc calo mvh matu , by(gender)

The repeated sampling of participants resulted in a hierarchical data set that consisted of within (level 1) and between (level 2) measurement variations. Table 1 outlines the variables used in the analysis, their means, and their standard deviations within and between participants. Shapiro-Wilk's tests did not reject the null hypothesis that the radial and tibial bone property distributions are normally distribution. Radial SOS measured an average of 129 points greater than tibial SOS (p=0.00). Girls, on average, has greater bone property measurements than boys; radial SOS by 19 (p=0.0495) and tibial SOS 22 points (0.0376). 

The bone property variances exhibited two distinct characteristics. One is that notable between and within variances existed. Another is that greater variation was exhibited across participants (i.e., between) than across time (i.e., within). The standard deviation of radial SOS measurements across participants (i.e., between) was 86 m/s compared to 51 m/s. The across and with participant standard deviation of tibial SOS was 96 and 50 m/s, respectively. Each independent variable also had a greater between compared to within variance to suggest that a fixed-participant estimation procedure may lead to considerable efficiency loss and a random-participant model is appropriate.


A multilevel structural equation model (i.e., SEM) was used to describe bone properties changes as postulated by the functional model of bone development. According to this model, specific modulators influence muscle strength, which in turn, along with other cellular modulators related to osteoblasts/osteoclasts, influence bone properties (see Figure 1). 
Models were constructed using Stata 14’s Generalized Structural Equation package (see Kaplan, 2008; Kline, 2015 for an overview of structural equation modeling). In brief, SEM is a multivariate statistical technique used to estimate a system of equations and test hypotheses about the relationships among variables. To do so, SEM explicates the direct relationships between observed variables and the covariance relationships between unobserved variables. 


A multilevel structural equation model (i.e., SEM) was used to describe bone properties changes as postulated by the functional model of bone development. According to this model, specific modulators influence muscle strength, which in turn, along with other cellular modulators related to osteoblasts/osteoclasts, influence bone properties (see Figure 1). Models were constructed using Stata 14’s Generalized Structural Equation package (see Kaplan, 2008; Kline, 2015 for an overview of structural equation modeling). In brief, SEM is a multivariate statistical technique used to estimate a system of equations and test hypotheses about the relationships among variables. To do so, SEM explicates the direct relationships between observed variables and the covariance relationships between unobserved variables. 

In [ ]:
%%stata --data dm 
set processors 20
xtset id 

tab gender

gen male = (gender=="boy")

*sum  rsos tsos grip biodex ntxc calo matu mvh male
xtsum rsos tsos grip biodex ntxc calo matu mvh male





In [ ]:
%%stata --data dm 
set processors 20
xtset id 

xtsum rsos
xtreg rsos, re mle

*xtivreg rsos ntx matu (grip = matu cal), fe
loneway rsos id


In [62]:
%%stata --data dm

swilk rsos tsos

display "*******************************"
display "t-test radial SOS across gender"
ttest rsos, by(gender)

display "t-test tibial SOS across gender"
ttest tsos, by(gender)

ttest rsos==tsos


                   Shapiro-Wilk W test for normal data

    Variable |        Obs       W           V         z       Prob>z
-------------+------------------------------------------------------
        rsos |        406    0.99592      1.137     0.306    0.37983
        tsos |        406    0.99706      0.821    -0.469    0.68033
*******************************
t-test radial SOS across gender

Two-sample t test with equal variances
------------------------------------------------------------------------------
Variable |     Obs        Mean    Std. Err.   Std. Dev.   [95% Conf. Interval]
---------+--------------------------------------------------------------------
     boy |     209    3806.622    6.706162    96.94986    3793.401    3819.843
    girl |     197    3826.122    7.299002    102.4464    3811.727    3840.516
---------+--------------------------------------------------------------------
combined |     406    3816.084    4.963334    100.0084    3806.327    3825.841
---------+

In [ ]:
%%stata --data dm

gen lngrip = log(grip)
gen lnbiodex = log(biodex)


swilk grip biodex 
swilk lngrip lnbiodex

swilk matu ntxc calo mvh


In [ ]:
%%stata --data dm --graph

gen lngrip = log(grip)
gen lnbiodex = log(biodex)

histogram lnbiodex

In [ ]:
%%stata --data dm --graph

twoway (hist rsos, frac lcolor(gs12) fcolor(gs12)) (hist tsos, frac fcolor(none) lcolor(red)), legend(off) xtitle("rsos (red: tsos)")


In [ ]:
%%stata --data dm 
set processors 20
xtset id trip

xtsum rsos tsos grip biodex

xttab rsos

In [ ]:
dr = dm[['id', 'trip', 'rsos', 'grip', 'matu', 'ntxc', 'calo']]
dr = dr.set_index(['id', 'trip'])

d1 = dr.shift(1)
# dr['lnrsos'] = numpy.log(dr.rsos)
# # dr['r1'] = dr.lnrsos.shift(1)
# dr['rp'] = dr.lnrsos - dr.lnrsos.shift(1)

# dr['rsos1'] = dr.rsos.shift(1)
# dr['rsosd'] = (dr.rsos - dr.rsos1) 
# dr['rsosp'] = dr.rsosd / dr.rsos1
# dr['rsosp1'] = dr.rsos / dr.rsos.shift(1) -1

dp = dr / d1 - 1
# dp.groupby('trip').mean()

In [ ]:
rsos grip matu ntxc

In [ ]:
%%stata --data dm 
set processors 20
xtset id trip



ivregress 2sls st_rsos  st_ntx (st_grip = st_height st_weight st_cal ), first 
estat endogenous



In [ ]:
%%stata --data dm 
set processors 20
xtset id trip


display "Mediated"
sem (st_rsos <- st_grip st_matu st_ntx) (st_grip <- st_matu) , nolog nocapslatent

In [ ]:
%%stata --data dm 
set processors 20
xtset id trip

display "******************************************************************"
sureg (rsos grip matu ntxc) (tsos ptiso matu ntxc mvh), isure

display "******************************************************************"
sureg (rsos grip matu ntxc) (tsos biodex matu ntxc mvh), isure

display "******************************************************************"
sureg (st_rsos st_grip st_matu st_ntxc) (st_tsos st_ptiso st_matu st_ntxc st_mvh), isure

display "******************************************************************"
sureg (st_rsos st_grip st_matu st_ntxc) (st_tsos st_bio st_matu st_ntxc st_mvh), isure



In [13]:
%%stata --data dm 
set processors 20
xtset id trip

sem ///
(rad <- grp mat ntx) ///
(tib <- tor mat ntx mvh) /// 
, covstruct(_lexogenous, diagonal) nocapslatent ///
cov(e.rad*e.tib) 
    

display "Mediated"
sem ///
(rad <- grp mat ntx) (grp <- mat cal) ///
(tib <- tor mat ntx mvh)  (tor <- mat cal)  /// 
, covstruct(_lexogenous, diagonal) nocapslatent ///
cov(e.rad*e.tib) 


    The maximum number of processors or cores being used is changed from 32 to       panel variable:  id (unbalanced)
        time variable:  trip, 1 to 4
                delta:  1 unit
> (rad <- grp mat ntx) ///
> (tib <- tor mat ntx mvh) /// 
> , covstruct(_lexogenous, diagonal) nocapslatent ///
> cov(e.rad*e.tib) 
variable rad not found;
Perhaps you meant 'rad' to specify a latent variable.
For 'rad' to be a valid latent variable specification, 'rad' must appear in the latent() option.
r(111);

end of do-file

r(111);


In [ ]:
%%stata --data dm 
set processors 20
xtset id trip

gsem ///
(rad <- grp mat ntx M[id]@1) ///
(tib <- tor mat ntx mvh M[id]@1) /// 
, covstruct(_lexogenous, diagonal) latent(M) ///
cov(e.rad*e.tib)


display "Mediated"
gsem ///
(rad <- grp mat ntx M[id]@1) (grp <- mat cal) ///
(tib <- tor mat ntx mvh M[id]@1) (tor <- mat cal)  /// 
, covstruct(_lexogenous, diagonal) nocapslatent  latent(M) ///
cov(e.rad*e.tib) 

display "**********"

gsem ///
(rad <- grp mat ntx M[id]@1) (grp <- mat cal) ///
(tib <- tor mat ntx mvh M[id]@1) (tor <- mat cal)  /// 
, covstruct(_lexogenous, diagonal) nocapslatent  latent(M) ///
cov(e.rad*e.tib) cov(e.grp*e.tor)


In [ ]:
%%stata --data dm 
set processors 20
xtset id trip

gsem ///
(rad <- grp mat ntx M[id]@1) (grp <- mat cal) ///
(tib <- tor mat ntx mvh M[id]@1) (tor <- mat cal)  /// 
, covstruct(_lexogenous, diagonal) nocapslatent  latent(M) ///
cov(e.rad*e.tib) cov(e.grp*e.tor)   

lincom _b[grp:mat] - _b[tor:mat]
lincom _b[grp:cal] - _b[tor:cal]

lincom _b[rad:grp] - _b[tib:tor]
lincom _b[rad:mat] - _b[tib:mat]
lincom _b[rad:ntx] - _b[tib:ntx]

In [ ]:
gsem ///
    (M1[id] griphs matus ntxhs godin  -> rsoshs, ) /// 
    (M2[id] torqhs matus ntxhs mvhhs  -> tsoshs, ) /// 
    (cinths matus  -> griphs,)  ///
    (cinths matus  -> torqhs,)  ///
     , covstruct(_lexogenous, diagonal) nocapslatent ///
      latent(M1 M2)   ///
      cov(e.rsoshs*e.tsoshs)  cov(e.torqhs*e.griphs)  ///
      cov(e.rsoshs*e.griphs)  cov(e.tsoshs*e.torqhs) ///
      cov(M1[id]*M2[id])  
        
estimates save "./out/table2_b"

lincom [griphs]matus - [torqhs]matus
lincom [griphs]cinths - [torqhs]cinths

lincom [rsoshs]griphs - [tsoshs]torqhs
lincom [rsoshs]matus - [tsoshs]matus
lincom [rsoshs]ntxhs - [tsoshs]ntxhs
lincom [rsoshs]paqps - [tsoshs]mvhhs